Sources/NIOSSH/TransportProtection/SSHTransportProtection.swift

```diff
@@ -87,10 +87,10 @@ protocol NIOSSHTransportProtection: AnyObject {
     /// length, the padding, or the MAC), and update source to indicate the consumed bytes.
     /// It must also perform any integrity checking that
     /// is required and throw if the integrity check fails.
-    func decryptAndVerifyRemainingPacket(_ source: inout ByteBuffer) throws -> ByteBuffer
+    func decryptAndVerifyRemainingPacket(_ source: inout ByteBuffer, sequenceNumber: UInt32) throws -> ByteBuffer
```

Comment:
Actually, two additional questions before we make this API public:
1. Why do we return a `ByteBuffer` here instead of in-place mutation?
2. There is a bit of a lost symmetry here, in plaintext mode padding and packet length are consumed by parser, but in ciphertext mode we have to do it manually, should this be handled by parser instead? And decryptors should only decrypt data in place and leave it? (Maybe we should drop mac bytes here?)

Reply:
I think parser can handle mac bytes, we have `macBytes` property, we can use it to remove last bytes from the packet

Reply:
1. As the doc comment says: it is expected to update the buffer to indicate what it consumed, and return the decrypted buffer. We avoid in-place mutation because we do not want to guarantee that all encryption algorithms have the exact same size input and output.
2. The absence of symmetry is because in the input case we need to be able to decrypt just the length so we know how much further data to read, and we wanted to minimise the number of times we need to do that operation. I don't think the parser _can_ do it for us because the actual length is affected by the protection scheme.

We could pass the MAC separately, but I'm not sure we gain much do we?

Reply:
let me think about this further, I'll revisit this together with refactoring `encryptPacket`

Reply:
Just for my clarification, did this work get done?

Reply:
I opened a draft here:
https://github.com/artemredkin/swift-nio-ssh/pull/1
but so far I only refactored `encryptPacket`, we probably should have two separate PRs for `encryptPacket` and `decryptAndVerifyRemainingPacket`

Reply:
Ok great, so you'd like to have this PR merged without further changes to this and we'll revisit in a follow-on PR?

Reply:
yes! otherwise it's already quite large